pkg/apis/config/v1alpha1/types.go

```diff
@@ -0,0 +1,276 @@
+type TemplatePolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Policy template.Config `json:"policy"`
+	// +optional
+	Common CommonConfig `json:"common,omitempty"`
+	Status ConfigStatus `json:"status,omitempty"`
+}
+
+// TemplatePolicyList represents a list of TemplatePolicies.
+// +kubebuilder:object:root=true
+type TemplatePolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []TemplatePolicy `json:"items"`
+}
+
+type CommonConfig struct {
+	// +optional
+	Control control.Config `json:"control,omitempty"`
+	// +optional
+	Log log.Config `json:"log,omitempty"`
+	// +optional
+	Instrumentation instrumentation.Config `json:"instrumentation,omitempty"`
+}
+
+// ConfigStatus is the per-node status for a configuration resource.
+type ConfigStatus struct {
+	Nodes map[string]NodeStatus `json:"nodes"`
+}
+
+// NodeStatus is the configuration status for a single node.
+type NodeStatus struct {
+	// Status of activating the configuration on this node.
+	// +kubebuilder:validation:Enum=Success;Failure
+	Status string `json:"status"`
+	// Generation is the generation the configuration this status was set for.
+	Generation int64 `json:"generation"`
+	// Error can provide further details of a configuration error.
+	Error *string `json:"errors,omitempty"`
+	// Timestamp of setting this status.
+	Timestamp metav1.Time `json:"timestamp,omitempty"`
+}
+
+const (
+	StatusSuccess = metav1.StatusSuccess
+	StatusFailure = metav1.StatusFailure
+)
+
+// NewNodeStatus returns a node status for the given error.
+func NewNodeStatus(err error, generation int64) *NodeStatus {
+	s := &NodeStatus{
+		Generation: generation,
+		Timestamp:  metav1.Now(),
+	}
+	if err == nil {
+		s.Status = StatusSuccess
+		// TODO(klihub): 'Patch away' any old errors from lingering. I don't
+		//     know if there is a nicer way of doing this with Patch().
+		e := ""
+		s.Error = &e
+	} else {
+		s.Status = StatusFailure
+		e := fmt.Sprintf("%v", err)
+		s.Error = &e
+	}
+	return s
+}
+
+type patchConfig struct {
+	Status patchStatus `json:"status,omitempty"`
+}
+
+type patchStatus struct {
+	Nodes map[string]*NodeStatus `json:"nodes,omitempty"`
+}
+
+func NodeStatusPatch(node string, status *NodeStatus) ([]byte, types.PatchType, error) {
+	cfg := &patchConfig{
+		Status: patchStatus{
+			Nodes: map[string]*NodeStatus{
+				node: status,
+			},
+		},
+	}
+
+	data, err := json.Marshal(cfg)
+	if err != nil {
+		return nil, types.PatchType(""), fmt.Errorf("failed to marshal patch: %v", err)
+	}
+
+	return data, types.MergePatchType, nil
+}
+
+// AnyConfig provides access to metadata of configuration resources.
+// All top level configuration types must implement this interface.
+// Among other things, the agent uses it to filter out unnecessary
+// update notifications.
+// +kubebuilder:object:generate=false
+type AnyConfig interface {
+	GetObjectMeta() *metav1.ObjectMeta
+}
```

Comment:
You shouldn't need this. E.g. k8s apimachinery (`k8s.io/apimachinery/pkg/apis/meta/v1`) provides helpers - you should be able to use the generic `Object` type (`metav1.Object`) from there to access object meta fields of any api object. The k8s `client-go` provides some similar functionalities iirc.

```suggestion
```

Reply:
A sidenote to this that I think you need to inform the kubebuilder to implement the interface with
```
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```
in the type definition

Reply:
Removed and switched to using `metav1.ObjectMeta` and `metav1.ObjectMetaAccessor` type assertions.